packages/integration-objects/srv/integration-objects/simple-exchange-rate-type-service.js

```diff
@@ -0,0 +1,210 @@
+    await checkUniquenessForPrimaryKeys(req);
+  }
+
+  function beforeDelete(req) {
+    try {
+      validateDelete(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function beforeCreateForRateTypeDescriptions(req) {
+    try {
+      await validateDescPayload(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function beforeUpdateForRateTypeDescriptions(req) {
+    try {
+      await validateDescPayload(req);
+    } catch (err) {
+      throw new ValidationError(err.message, err.code);
+    }
+  }
+
+  async function validateDescPayload(req) {
+    validateFieldsForExchangeRateTypeDescription(req.data, req.params);
+    await checkForValidityOfKey(req, ExchangeRateTypes);
+    await checkUniquenessForPrimaryKeysForExchangeRateTypeDescForUpdate(req);
+  }
+
+  async function checkUniquenessForPrimaryKeys(req) {
+    const record = req.data;
+    const affectedRows = await fetchExistingRateTypes(ExchangeRateTypes, record, req);
+    checkForDuplicateRateTypeObjects(affectedRows, req);
+  }
+
+  async function checkUniquenessForPrimaryKeysForExchangeRateTypeDescForUpdate(req) {
+    const record = req.data;
+    const affectedRows = await fetchExistingRateTypeDesc(ExchangeRateTypeDescriptions, record, req);
+    checkForDuplicateRateTypeObjects(affectedRows, req);
+  }
+
+  async function beforeActiveDeleteForExchangeRateTypeDescription(req) {
+    const record = req.data;
+    let id;
+    const affectedRows = await fetchExistingTexts(ExchangeRateTypeDescriptions, record, req);
+    if (affectedRows.length) {
+      id = affectedRows[0].ID_texts;
+    }
+    await validateIdInTenant(id, ExchangeRateTypes, req, affectedRows);
+  }
+
+  async function writeAuditLogForRateTypes(req, next) {
+    let currentValues;
+    const data = req.data;
+    const type = 'exchange-rate-type';
+    const auditParams = {
+      exchangeRateType: data.exchangeRateType,
+      exchangeRateTypeDescription: data.exchangeRateTypeDescription,
+      referenceCurrencyThreeLetterISOCode: data.referenceCurrencyThreeLetterISOCode
+    };
+
+    let currentValuesIsNotEmpty = true;
+    if (Constants.CREATE_EVENT !== req.event) {
+      currentValues = await fetchCurrentAttributeValues(req, req.user.tenant, ExchangeRateTypes);
+      currentValuesIsNotEmpty = !util.isNullish(currentValues);
+    }
+
+    if (currentValuesIsNotEmpty) {
+      await writeAuditLog(req, auditParams, currentValues, type);
+    }
+    await next();
+  }
+};
+
+async function validateIdInTenant(id, ExchangeRateTypes, req, affectedRows) {
+  if (id != null) {
+    const query2 = SELECT.from(ExchangeRateTypes).where({ ID: id }).limit(1, 0);
+    const tx2 = cds.transaction(req);
+    const affectedRows2 = await tx2.run(query2);
+
+    if (affectedRows2.length) {
+      const record2 = affectedRows[0];
+      if (req.user.tenant !== record2.tenantID) {
```

Comment:
Why do those names end with a 2?
```suggestion
    const query = SELECT.from(ExchangeRateTypes).where({ ID: id }).limit(1, 0);
    const tx = cds.transaction(req);
    const firstRowForId = await tx.run(query);

    if (firstRowForId.length) {
      const [record] = affectedRows;
      if (req.user.tenant !== record.tenantID) {
```